Updated config loading to match #1

sjrct opened the pull request:
Some minor differences from #1:
- 'name' was dropped from the item config files as it is inferred
- 'srcs' array was replaced with the string 'main'

To load the module 'foo' the relevant paths are searched for directories
of the given name that contain a 'config.json'. This config file is used
as the config file for the module or agent (at a class level).

Sample config:

```
{
  "module-path": [
    "modules"
  ],

  "agent-path": [
    "agents"
  ],

  "agent-instances": {
    "irc0": {
      "of": "irc",
      "channel": "##halibot"
    }
  },

  "module-instances": {
    "quote0": {
      "of": "quote"
    }
  }
}
```

Resolves #1

sjrct committed:
Updated config loading to match #1

Some minor differences from #1:
  - 'name' was dropped from the item config files as it is inferred
  - 'srcs' array was replaced with the string 'main'

To load the module 'foo' the relevant paths are searched for directories
of the given name that contain a 'config.json'. This config file is used
as the config file for the module or agent (at a class level).

richteer commented:
From a quick skim, looks good. Eventually those additional functions might be better off in their own file, but not important now. Might be able to remove some of the old imports, again not critical.

sjrct commented:
Yeah, I did not cull the imports. I was thinking in the long run it might be useful to have a `Loader` class or something of the like. Merge for now though?

sjrct committed:
Implemented Loader abstraction

This abstraction absorbs the generalized module/agent loading functions
previously found in halibot.py and brings them into the more generalized
Loader class.

sjrct commented:
Err... Or I will just implement the `Loader` abstraction.

sjrct committed:
name on config -> name on agent/module

This seems like the more sensible choice, as now the config variable
contains exclusively the contents of the json config. Also shorter to
type on the agent/module implementer's end.
